### PR TITLE
Feature | Added support for Always Encrypted with Secure Enclaves

### PR DESCRIPTION
**Pull request by rene-ye**

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/AE.java [hunk 1]
```diff
@@ -27,6 +27,21 @@
         algorithmName = algorithmNameVal;
     }
 
+    public EncryptionKeyInfo(byte[] encryptedKeyVal, int dbId, int keyId, int keyVersion, byte[] mdVersion,
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/AE.java]: This doesn't need to be public.

File: src/main/java/com/microsoft/sqlserver/jdbc/AE.java [hunk 2]
```diff
@@ -114,6 +131,30 @@ void add(byte[] encryptedKey, int dbId, int keyId, int keyVersion, byte[] mdVers
             assert ((null != cekMdVersion) && (null != mdVersion) && (cekMdVersion.length == mdVersion.length));
         }
     }
+
+    public void add(byte[] encryptedKey, int dbId, int keyId, int keyVersion, byte[] mdVersion, String keyPath,
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/AE.java]: I can't figure out where this method is used. It also probably doesn't need to be public.

File: src/main/java/com/microsoft/sqlserver/jdbc/AE.java [hunk 3]
```diff
@@ -267,4 +309,28 @@ int value() {
         return ordinal() + 1;
     }
 
+    /**
+     * Fields in the third resultset of "sp_describe_parameter_encryption". We expect the server to return the fields in the
+     * resultset in the same order as mentioned below. If the server changes the below order, then transparent parameter
+     * encryption will break.
+     */
+    enum DescribeParameterEncryptionResultSet3 {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/AE.java]: Not used anywhere.

**Comment by ulvii** [comment]
Please format the files.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 4]
```diff
@@ -873,4 +873,19 @@
      *        boolean value for 'useFmtOnly'.
      */
     void setUseFmtOnly(boolean useFmtOnly);
+
+    /**
+     * Returns the enclave attestional url used with enclave based Always Encrypted.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     * Returns the enclave attestation URL used by Always Encrypted with Secure Enclaves.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 5]
```diff
@@ -873,4 +873,19 @@
...
+     * 
+     * @return a String that contains the enclave attestation url.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     * @return Enclave attestation URL.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 6]
```diff
@@ -873,4 +873,19 @@
...
+     */
+    String getEnclaveAttestationUrl();
+
+    /**
+     * Sets the enclave attestional url to be used with enclave based Always Encrypted.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     * Sets the enclave attestation URL used by Always Encrypted with Secure Enclaves.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 7]
```diff
@@ -873,4 +873,19 @@
...
+     * 
+     * @param url
+     *        A String that contains the enclave attestation url.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     *        Enclave attestation URL.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import java.security.InvalidAlgorithmParameterException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.util.ArrayList;
+
+
+public interface ISQLServerEnclaveProvider {
+    byte[] getEnclavePackage(String userSQL, ArrayList<byte[]> enclaveCEKs) throws SQLServerException;
+
+    void getAttestationParamters(boolean b,
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java]: Will all implementations of this interface throw the same exceptions? If not, this is implementation specific and needs to be handled in the class.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+class EnclaveSession {
+    private byte[] sessionID;
+    private long counter;
+    private byte[] sessionSecret;
+
+    // Don't allow default instantiation
+    @SuppressWarnings("unused")
+    private EnclaveSession() {};
+
+    public EnclaveSession(byte[] cs, byte[] b) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java]: remove public

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    byte[] getSessionID() {
+        return sessionID;
+    }
+
+    byte[] getSessionSecret() {
+        return sessionSecret;
+    }
+
+    long getCounter() {
+        counter++;
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java]: Please add a comment that explains this.

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java [hunk 11]
```diff
@@ -6189,6 +6191,26 @@ void writeRPCReaderUnicode(String sName, Reader re, long reLength, boolean bOut,
         // Write the data
         writeReader(re, reLength, usePLP);
     }
+
+    boolean isConnectionAEv2() {
+        return con.isAEv2();
+    }
+
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java]: why is this necessary?

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java (hunk 10 above)
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java]: perhaps a better name? typically getters don't change values

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java [hunk 12]
```diff
@@ -1544,13 +1544,14 @@ private TDSWriter sendBulkCopyCommand(TDSCommand command) throws SQLServerExcept
         // Create and send the initial command for bulk copy ("INSERT BULK ...").
         TDSWriter tdsWriter = command.startRequest(TDS.PKT_QUERY);
         String bulkCmd = createInsertBulkCommand(tdsWriter);
+        tdsWriter.sendEnclavePackage(null, null);
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java]: Is BulkCopy supposed to have this?

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java [hunk 13]
```diff
@@ -6189,6 +6193,22 @@ void writeRPCReaderUnicode(String sName, Reader re, long reLength, boolean bOut,
         // Write the data
         writeReader(re, reLength, usePLP);
     }
+
+    void sendEnclavePackage(String sql, ArrayList<byte[]> enclaveCEKs) throws SQLServerException {
+        if (con.isAEv2()) {
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java]: should null check this con

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
+package com.microsoft.sqlserver.jdbc;
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java]: add MIT licensing to file header

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 15]
```diff
@@ -873,4 +873,35 @@
      *        boolean value for 'useFmtOnly'.
      */
     void setUseFmtOnly(boolean useFmtOnly);
+
+    /**
+     * Returns the enclave attestation url used with enclave based Always Encrypted.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     * Returns the attestation URL used by Always Encrypted with Secure Enclaves.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 16]
```diff
@@ -873,4 +873,35 @@
...
+     * 
+     * @return Enclave attestation url.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     * @return enclave attestation URL.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 17]
```diff
@@ -873,4 +873,35 @@
...
+     */
+    String getEnclaveAttestationUrl();
+
+    /**
+     * Sets the enclave attestation url to be used with enclave based Always Encrypted.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     * Sets the enclave attestation URL used by Always Encrypted with Secure Enclaves.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 18]
```diff
@@ -873,4 +873,35 @@
...
+     * 
+     * @param url
+     *        Enclave attestation url.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]:
```suggestion
     *        Enclave attestation URL.
```

Comment (ulvii) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: url should be replaced by URL

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java [hunk 19]
```diff
@@ -0,0 +1,533 @@
+package com.microsoft.sqlserver.jdbc;
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java]: license

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 20]
```diff
@@ -873,4 +873,35 @@
...
+     */
+    void setEnclaveAttestationUrl(String url);
+
+    /**
+     * Returns the enclave attestation protocol used with enclave based Always Encrypted.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: The official name of the feature is "Always Encrypted with Secure Enclaves". I have not encountered the term "enclave based Always Encrypted" anywhere, please update all instances.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 21]
```diff
@@ -873,4 +873,35 @@
...
+     * 
+     * @return Enclave attestation protocol.
+     */
+    String getEnclaveAttestationProtocol();
+
+    /**
+     * Sets the enclave attestation protocol to be used with enclave based Always Encrypted.
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: Always Encrypted with Secure Enclaves

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java [hunk 22]
```diff
@@ -0,0 +1,533 @@
...
+
+import javax.crypto.KeyAgreement;
+
+
+public class SQLServerVSMEnclaveProvider implements ISQLServerEnclaveProvider {
+
+    private VSMAttestationParameters vsmParams = null;
+    private AttestationResponse hgsResponse = null;
+    private String attestationURL = null;
+    EnclaveSession enclaveSession = null;
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java]: private

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java [hunk 23]
```diff
@@ -0,0 +1,533 @@
...
+                        ByteBuffer aev2CekEntry = ByteBuffer.allocate(46);
+                        aev2CekEntry.order(ByteOrder.LITTLE_ENDIAN).putInt(dbID);
+                        aev2CekEntry.put(mdVer);
+                        aev2CekEntry.putShort((short) keyID);
+                        aev2CekEntry.put(connection.getColumnEncryptionKeyStoreProvider(keyStoreName)
+                                .decryptColumnEncryptionKey(keyPath, algo, encryptedKey));
+                        enclaveRequestedCEKs.add(aev2CekEntry.array());
+                    }
+                }
+                // if (getStatementLogger().isLoggable(java.util.logging.Level.FINE)) {
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java]: remove comment

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionKeyStoreProvider.java [hunk 24]
```diff
@@ -62,4 +62,19 @@
     public abstract byte[] encryptColumnEncryptionKey(String masterKeyPath, String encryptionAlgorithm,
             byte[] columnEncryptionKey) throws SQLServerException;
 
+    /**
+     * Verify the signature is valid for the column master key
+     * 
+     * @param masterKeyPath
+     *        column master key path
+     * @param allowEnclaveComputations
+     *        indicates if whether the column master key supports enclave computations
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionKeyStoreProvider.java]:
```suggestion
     *        indicates whether the column master key supports enclave computations.
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 25]
```diff
@@ -1422,6 +1464,36 @@ Connection connectInternal(Properties propsIn,
...
+                    if (connectionlogger.isLoggable(Level.SEVERE)) {
+                        connectionlogger.severe(toString() + " "
+                                + SQLServerException.getErrString("R_enclaveInvalidAttestationProtocol"));
+                    }
+                    throw new SQLServerException(SQLServerException.getErrString("R_enclaveInvalidAttestationProtocol"),
+                            null);
+                }
+            }
+
+            // enclaveAttestationUrl but no enclaveAttestationProtocol specified
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: What if enclaveAttestationProtocol specified, but there is no enclaveAttestationUrl ?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java [hunk 26]
```diff
@@ -113,6 +113,46 @@ static ColumnEncryptionSetting valueOfString(String value) throws SQLServerExcep
 }
 
 
+enum AttestationProtocol {
+    HGS("HGS"); // only protocol supported currently
+
+    private final String protocol;
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java]: this field along with EnclaveType.type seems like they're never used.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 27]
```diff
@@ -4553,13 +4633,33 @@ private void onFeatureExtAck(byte featureId, byte[] data) throws SQLServerExcept
...
+
+                if (null != enclaveAttestationUrl) {
+                    if (aeVersion < TDS.COLUMNENCRYPTION_VERSION2) {
+                        throw new SQLServerException(SQLServerException.getErrString("R_enclaveNotSupported"), null);
+                    } else {
+                        serverColumnEncryptionVersion = ColumnEncryptionVersion.AE_v2;
+                        enclaveType = new String(data, 2, data.length - 2, UTF_16LE);
+                    }
+
+                    if (null == enclaveType) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: I don't think this check makes sense, `new String()` will set  `enclaveType` to a non-null value. Better perform the check on `data`.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java (hunk 19 above)
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java]: license here

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java [hunk 28]
```diff
@@ -0,0 +1,533 @@
...
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.crypto.KeyAgreement;
+
+
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java]:
```suggestion
@SuppressWarnings("unused")
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java [hunk 29]
```diff
@@ -0,0 +1,533 @@
...
+        KeyAgreement ka = KeyAgreement.getInstance("ECDH");
+        ka.init(privateKey);
+        // Generate a PublicKey from the above key specifications and do an agreement with our PrivateKey
+        ka.doPhase(KeyFactory.getInstance("EC").generatePublic(keySpec), true);
+        // Generate a Secret from the agreement and hash with SHA-256 to create Session Secret
+        return MessageDigest.getInstance("SHA-256").digest(ka.generateSecret());
+    }
+}
+
+
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerVSMEnclaveProvider.java]:
```suggestion
@SuppressWarnings("unused")
```

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java [hunk 30]
```diff
@@ -72,24 +139,92 @@
...
+        String enclaveAttestationProtocol = TestUtils.getConfiguredProperty("enclaveAttestationProtocol");
+        if (null != enclaveAttestationProtocol) {
+            tmpConnectionString = TestUtils.addOrOverrideProperty(tmpConnectionString, "enclaveAttestationProtocol",
+                    enclaveAttestationProtocol);
+        }
+
+        // add enclave properties if AEv2 supported
+        try (Connection con = PrepUtil.getConnection(tmpConnectionString)) {
+            if (TestUtils.isAEv2(con)) {
+                isAEv2Supported = true;
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java]: This is not required anymore.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java [hunk 31]
```diff
@@ -54,14 +59,76 @@
...
-    static String keyPath = null;
+    static String javaKeyPath = null;
     static String javaKeyAliases = null;
-
-    static SQLServerColumnEncryptionKeyStoreProvider storeProvider = null;
+    static String windowsKeyPath = null;
+    static String applicationClientID = null;
+    static String applicationKey = null;
+    static String[] keyIDs = null;
+    static String cmk_jks = null;
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java]: Please name these camel case.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java [hunk 32]
```diff
@@ -0,0 +1,54 @@
+package com.microsoft.sqlserver.jdbc;
+
+import java.security.PrivateKey;
+import java.util.ArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+public interface ISQLServerEnclaveProvider {
+    byte[] getEnclavePackage(String userSQL, ArrayList<byte[]> enclaveCEKs) throws SQLServerException;
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java]: We should add javadoc to these methods, since they're public APIs.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java [hunk 33]
```diff
@@ -712,10 +538,10 @@ private static void createCEK(SQLServerColumnEncryptionKeyStoreProvider storePro
      * @throws SQLException
      */
     protected static void dropTables(Statement stmt) throws SQLException {
-        TestUtils.dropTableIfExists(AbstractSQLGenerator.escapeIdentifier(Constants.NUMERIC_TABLE_AE), stmt);
-        TestUtils.dropTableIfExists(AbstractSQLGenerator.escapeIdentifier(Constants.CHAR_TABLE_AE), stmt);
-        TestUtils.dropTableIfExists(AbstractSQLGenerator.escapeIdentifier(Constants.BINARY_TABLE_AE), stmt);
-        TestUtils.dropTableIfExists(AbstractSQLGenerator.escapeIdentifier(Constants.DATE_TABLE_AE), stmt);
+        TestUtils.dropTableIfExists(AbstractSQLGenerator.escapeIdentifier(NUMERIC_TABLE_AE), stmt);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java]: I thought we were keeping all constants in one file?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 34]
```diff
@@ -594,5 +596,27 @@ static String getResource(String key) {
...
+            {"R_enclaveNotSupported", "The SQL Server instance does not support enclave based computations."},
+            {"R_enclaveNoAttestationProtocol",
+                    "The \"enclaveAttestationProtocol\" connection property must be specified with \"enclaveAttestationUrl\"."},
+            {"R_enclaveInvalidAttestationProtocol", "The \"enclaveAttestationProtocol\" is invalid."},
+            {"R_enclaveTypeNotReturned",
+                    "The SQL Server instance did not return an enclave type. Please make sure the enclave type is correct configured in the SQL Server instance."},
+            {"R_enclaveTypeInvalid", "The enclave type {0} is invalid or not supported by the driver."},
+            {"R_attestationUrlInvalid", "Unable to attest enclave specified by {0}."},
+            {"R_EnclaveResponseLengthError",
+                    "More bytes from the server was recieved than expected when parsing Enclave Attestation response."},
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java]:
```suggestion
                    "More bytes from the server were received than expected when parsing Enclave Attestation response."},
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 35]
```diff
@@ -594,5 +596,27 @@ static String getResource(String key) {
...
+            {"R_EnclavePackageLengthError",
+                    "More bytes from the server was recieved than expected when parsing Enclave Package."},
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java]:
```suggestion
                    "More bytes from the server were received than expected when parsing Enclave Package."},
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 36]
```diff
@@ -594,5 +596,27 @@ static String getResource(String key) {
...
+            {"R_EnclavePKLengthError",
+                    "More bytes from the server was recieved than expected when parsing Enclave PK."},
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java]:
```suggestion
                    "More bytes from the server were received than expected when parsing Enclave PK."},
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java (hunk 27 above)
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: data is not null
